File: android-app/app/src/test/java/com/summerveldhoundresort/app/db/entities/EventTest.kt
package com.summerveldhoundresort.app.db.entities

import org.junit.Test
import org.junit.Assert.*

class EventTest {

    @Test
    fun `event creation with all parameters should work correctly`() {
        // Given
        val eventId = "event123"
        val name = "Dog Training Workshop"
        val description = "Learn basic obedience training for your dog"
        val date = "2024-02-15"
        val location = "Main Training Hall"
        val time = "10:00 AM"

        // When
        val event = Event(eventId, name, description, date, location, time)

        // Then
        assertEquals(eventId, event.id)
        assertEquals(name, event.name)
        assertEquals(description, event.description)
        assertEquals(date, event.date)
        assertEquals(location, event.location)
        assertEquals(time, event.time)
    }

    @Test
    fun `event creation with empty constructor should work`() {
        // When
        val event = Event()

        // Then
        assertTrue(event.id.isEmpty())
        assertTrue(event.name.isEmpty())
        assertTrue(event.description.isEmpty())
        assertTrue(event.date.isEmpty())
        assertTrue(event.location.isEmpty())
        assertTrue(event.time.isEmpty())
    }

    @Test
    fun `event with special characters in name should work`() {
        // Given
        val specialName = "Dog's Day Out & Training Session"
        val event = Event(
            id = "event123",
            name = specialName,
            description = "Special event with special characters",
            date = "2024-02-15",
            location = "Main Hall",
            time = "10:00 AM"
        )

        // Then
        assertEquals(specialName, event.name)
    }

    @Test
    fun `event with long description should work`() {
        // Given
        val longDescription = "This is a very long description for an event that contains " +
                "multiple sentences and provides detailed information about what will " +
                "happen during the event, including all the activities and requirements."
        val event = Event(
            id = "event123",
            name = "Test Event",
            description = longDescription,
            date = "2024-02-15",
            location = "Main Hall",
            time = "10:00 AM"
        )

        // Then
        assertEquals(longDescription, event.description)
        assertTrue(event.description.length > 100)
    }

    @Test
    fun `event with different date formats should work`() {
        // Given
        val dateFormats = listOf(
            "2024-02-15",
            "15/02/2024",
            "February 15, 2024",
            "2024-02-15T10:00:00Z"
        )

        dateFormats.forEach { dateFormat ->
            // When
            val event = Event(
                id = "event123",
                name = "Test Event",
                description = "Test description",
                date = dateFormat,
                location = "Main Hall",
                time = "10:00 AM"
            )

            // Then
            assertEquals(dateFormat, event.date)
        }
    }

    @Test
    fun `event with different time formats should work`() {
        // Given
        val timeFormats = listOf(
            "10:00 AM",
            "10:00",
            "22:00",
            "10:00:00",
            "10:00 AM - 12:00 PM"
        )

        timeFormats.forEach { timeFormat ->
            // When
            val event = Event(
                id = "event123",
                name = "Test Event",
                description = "Test description",
                date = "2024-02-15",
                location = "Main Hall",
                time = timeFormat
            )

            // Then
            assertEquals(timeFormat, event.time)
        }
    }

    @Test
    fun `event with empty strings should work`() {
        // Given
        val event = Event(
            id = "",
            name = "",
            description = "",
            date = "",
            location = "",
            time = ""
        )

        // Then
        assertTrue(event.id.isEmpty())
        assertTrue(event.name.isEmpty())
        assertTrue(event.description.isEmpty())
        assertTrue(event.date.isEmpty())
        assertTrue(event.location.isEmpty())
        assertTrue(event.time.isEmpty())
    }

    @Test
    fun `event with null-like values should work`() {
        // Given
        val event = Event(
            id = "null",
            name = "null",
            description = "null",
            date = "null",
            location = "null",
            time = "null"
        )

        // Then
        assertEquals("null", event.id)
        assertEquals("null", event.name)
        assertEquals("null", event.description)
        assertEquals("null", event.date)
        assertEquals("null", event.location)
        assertEquals("null", event.time)
    }
}

File: backend/tests/event.test.js
// Mock Firebase Admin SDK
jest.mock('../config/firebase', () => ({
  initializeFirebase: jest.fn(() => ({
    firestore: jest.fn(() => ({
      collection: jest.fn(() => ({
        add: jest.fn(),
        where: jest.fn(() => ({
          limit: jest.fn(() => ({
            get: jest.fn()
          }))
        })),
        doc: jest.fn(() => ({
          get: jest.fn(),
          update: jest.fn(),
          delete: jest.fn()
        }))
      }))
    }))
  }))
}))

describe('Event Management', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Event Data Models', () => {
    it('should create event with all required fields', () => {
      const eventData = {
        id: 'event123',
        name: 'Dog Training Workshop',
        description: 'Learn basic obedience training for your dog',
        date: '2024-02-15',
        location: 'Main Training Hall',
        time: '10:00 AM'
      }

      expect(eventData.id).toBe('event123')
      expect(eventData.name).toBe('Dog Training Workshop')
      expect(eventData.description).toBe('Learn basic obedience training for your dog')
      expect(eventData.date).toBe('2024-02-15')
      expect(eventData.location).toBe('Main Training Hall')
      expect(eventData.time).toBe('10:00 AM')
    })

    it('should validate event name is not empty', () => {
      const eventData = {
        name: '',
        description: 'Test description',
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      expect(eventData.name).toBe('')
      expect(eventData.name.length).toBe(0)
    })

    it('should validate event date format', () => {
      const validDates = [
        '2024-02-15',
        '2024-12-31',
        '2025-01-01'
      ]

      validDates.forEach(date => {
        expect(date).toMatch(/^\d{4}-\d{2}-\d{2}$/)
      })
    })

    it('should validate event time format', () => {
      const validTimes = [
        '10:00 AM',
        '2:30 PM',
        '09:15',
        '14:30'
      ]

      validTimes.forEach(time => {
        expect(time).toBeTruthy()
        expect(time.length).toBeGreaterThan(0)
      })
    })
  })

  describe('Event Creation', () => {
    it('should create event with valid data', () => {
      const eventData = {
        name: 'Pet Grooming Session',
        description: 'Professional grooming for your pet',
        date: '2024-02-20',
        location: 'Grooming Room',
        time: '2:00 PM'
      }

      // Validate all required fields are present
      expect(eventData.name).toBeDefined()
      expect(eventData.description).toBeDefined()
      expect(eventData.date).toBeDefined()
      expect(eventData.location).toBeDefined()
      expect(eventData.time).toBeDefined()
    })

    it('should handle special characters in event name', () => {
      const eventData = {
        name: "Dog's Day Out & Training Session",
        description: 'Special event with special characters',
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      expect(eventData.name).toContain("'")
      expect(eventData.name).toContain('&')
    })

    it('should handle long descriptions', () => {
      const longDescription = 'This is a very long description for an event that contains ' +
        'multiple sentences and provides detailed information about what will ' +
        'happen during the event, including all the activities and requirements.'

      const eventData = {
        name: 'Test Event',
        description: longDescription,
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      expect(eventData.description.length).toBeGreaterThan(100)
    })
  })

  describe('Event Validation', () => {
    it('should validate required fields', () => {
      const requiredFields = ['name', 'description', 'date', 'location', 'time']
      const eventData = {
        name: 'Test Event',
        description: 'Test description',
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      requiredFields.forEach(field => {
        expect(eventData[field]).toBeDefined()
        expect(eventData[field]).not.toBe('')
      })
    })

    it('should validate date is not in the past', () => {
      const today = new Date()
      const tomorrow = new Date(today)
      tomorrow.setDate(tomorrow.getDate() + 1)

      const eventDate = tomorrow.toISOString().split('T')[0]

      expect(eventDate).toBeDefined()
      expect(new Date(eventDate)).toBeInstanceOf(Date)
    })

    it('should validate time format', () => {
      const validTimeFormats = [
        '10:00 AM',
        '2:30 PM',
        '09:15',
        '14:30',
        '10:00 AM - 12:00 PM'
      ]

      validTimeFormats.forEach(time => {
        expect(time).toBeTruthy()
        expect(typeof time).toBe('string')
      })
    })
  })

  describe('Event Search and Filtering', () => {
    it('should filter events by name', () => {
      const events = [
        { id: '1', name: 'Dog Training', description: 'Basic training', date: '2024-02-15', location: 'Main Hall', time: '10:00 AM' },
        { id: '2', name: 'Pet Grooming', description: 'Professional grooming', date: '2024-02-16', location: 'Grooming Room', time: '2:00 PM' },
        { id: '3', name: 'Advanced Training', description: 'Advanced obedience', date: '2024-02-17', location: 'Training Room', time: '3:00 PM' }
      ]

      const searchQuery = 'training'
      const filteredEvents = events.filter(event =>
        event.name.toLowerCase().includes(searchQuery.toLowerCase())
      )

      expect(filteredEvents).toHaveLength(2)
      expect(filteredEvents[0].name).toContain('Training')
      expect(filteredEvents[1].name).toContain('Training')
    })

    it('should filter events by date', () => {
      const events = [
        { id: '1', name: 'Event 1', description: 'Description 1', date: '2024-02-15', location: 'Main Hall', time: '10:00 AM' },
        { id: '2', name: 'Event 2', description: 'Description 2', date: '2024-02-16', location: 'Main Hall', time: '2:00 PM' },
        { id: '3', name: 'Event 3', description: 'Description 3', date: '2024-02-15', location: 'Training Room', time: '3:00 PM' }
      ]

      const targetDate = '2024-02-15'
      const eventsForDate = events.filter(event => event.date === targetDate)

      expect(eventsForDate).toHaveLength(2)
      expect(eventsForDate[0].date).toBe(targetDate)
      expect(eventsForDate[1].date).toBe(targetDate)
    })

    it('should filter events by location', () => {
      const events = [
        { id: '1', name: 'Event 1', description: 'Description 1', date: '2024-02-15', location: 'Main Hall', time: '10:00 AM' },
        { id: '2', name: 'Event 2', description: 'Description 2', date: '2024-02-16', location: 'Grooming Room', time: '2:00 PM' },
        { id: '3', name: 'Event 3', description: 'Description 3', date: '2024-02-17', location: 'Main Hall', time: '3:00 PM' }
      ]

      const targetLocation = 'Main Hall'
      const eventsAtLocation = events.filter(event => event.location === targetLocation)

      expect(eventsAtLocation).toHaveLength(2)
      expect(eventsAtLocation[0].location).toBe(targetLocation)
      expect(eventsAtLocation[1].location).toBe(targetLocation)
    })
  })

  describe('Event CRUD Operations', () => {
    it('should create new event', () => {
      const newEvent = {
        name: 'New Event',
        description: 'New event description',
        date: '2024-03-01',
        location: 'New Location',
        time: '11:00 AM'
      }

      expect(newEvent.name).toBe('New Event')
      expect(newEvent.description).toBe('New event description')
      expect(newEvent.date).toBe('2024-03-01')
      expect(newEvent.location).toBe('New Location')
      expect(newEvent.time).toBe('11:00 AM')
    })

    it('should update existing event', () => {
      const originalEvent = {
        id: 'event123',
        name: 'Original Event',
        description: 'Original description',
        date: '2024-02-15',
        location: 'Original Location',
        time: '10:00 AM'
      }

      const updatedEvent = {
        ...originalEvent,
        name: 'Updated Event',
        description: 'Updated description',
        time: '2:00 PM'
      }

      expect(updatedEvent.id).toBe(originalEvent.id)
      expect(updatedEvent.name).toBe('Updated Event')
      expect(updatedEvent.description).toBe('Updated description')
      expect(updatedEvent.date).toBe(originalEvent.date)
      expect(updatedEvent.location).toBe(originalEvent.location)
      expect(updatedEvent.time).toBe('2:00 PM')
    })

    it('should delete event', () => {
      const eventToDelete = {
        id: 'event123',
        name: 'Event to Delete',
        description: 'This event will be deleted',
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      expect(eventToDelete.id).toBe('event123')
      // In a real implementation, this would call a delete function
    })
  })

  describe('Event Error Handling', () => {
    it('should handle missing required fields', () => {
      const incompleteEvent = {
        name: 'Test Event'
        // Missing description, date, location, time
      }

      const requiredFields = ['description', 'date', 'location', 'time']
      const missingFields = requiredFields.filter(field => !incompleteEvent[field])

      expect(missingFields).toHaveLength(4)
      expect(missingFields).toEqual(['description', 'date', 'location', 'time'])
    })

    it('should handle invalid date format', () => {
      const invalidDates = [
        'invalid-date',
        '15/02/2024',
        '2024-13-01', // Invalid month
        '2024-02-30' // Invalid day
      ]

      invalidDates.forEach(date => {
        expect(date).not.toMatch(/^\d{4}-\d{2}-\d{2}$/)
      })
    })

    it('should handle empty event name', () => {
      const eventWithEmptyName = {
        name: '',
        description: 'Valid description',
        date: '2024-02-15',
        location: 'Main Hall',
        time: '10:00 AM'
      }

      expect(eventWithEmptyName.name).toBe('')
      expect(eventWithEmptyName.name.length).toBe(0)
    })
  })
})
